[ao][sparsity] L1 norm based block data sparsifier

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79535
* __->__ #79534
* #79254
* #79253
* #79427
* #79252
* #79251

L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Differential Revision: [D37164115](https://our.internmc.facebook.com/intern/diff/D37164115)

Commit:
[ao][sparsity] L1 norm based block data sparsifier

L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:
```diff
@@ -0,0 +1,139 @@
+        # Fetch the high level mask that zeros out entire blocks
+        data_lvl_mask = self._get_data_level_mask(data=data_norm, sparsity_level=sparsity_level,
+                                                  sparse_block_shape=sparse_block_shape)
+
```
```
if values_per_block == zeros_per_block:
  We are done
```

macandro96 replied:
done

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:
```diff
@@ -0,0 +1,139 @@
+        super().__init__(data_list=data_list, **defaults)
+
+    def _get_scatter_folded_mask(self, data, dim, indices, output_size, sparse_block_shape):
+        mask = torch.ones_like(data)
```
dtype should either bool or int or int8

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:
```diff
@@ -0,0 +1,139 @@
+
+        # create a new padded tensor like data (to match the block_shape)
+        padded_data = torch.ones(height + dh, width + dw, dtype=data.dtype, device=data.device)
+        padded_data = padded_data * float('inf')  # can also be replaced with 0 to stop the removal of edge data
```
try using `nan`

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:
```diff
@@ -0,0 +1,139 @@
+        mask = self._get_scatter_folded_mask(data=unfolded_data, dim=1, indices=sorted_idx, output_size=padded_data.shape,
+                                             sparse_block_shape=sparse_block_shape)
+
+        mask = mask.squeeze()[0:height, 0:width]  # remove padding
```
```
mask = mask.squeeze()[0:height, 0:width].contiguous()
```

macandro96 replied:
done

z-a-f reviewed (approved):
Some comments

HDCharles commented on test/ao/sparsity/test_data_sparsifier.py:
```diff
@@ -230,3 +232,209 @@ def test_nn_embeddings(self):
+        assert norm_type in ['L1']
+        self.norm_type = norm_type
+
+    def get_bounds_on_actual_sparsity(self, config, tensor_shape):
```
if these are helper classes for other tests, should probably prepend `_` to signify it being private.

HDCharles commented on test/ao/sparsity/test_data_sparsifier.py:
```diff
@@ -230,3 +232,209 @@ def test_nn_embeddings(self):
         emb_test.run_tests()
+
+
+class TestNormDataSparsifierType(TestBaseDataSparsiferType):
```
Similar to https://github.com/pytorch/pytorch/pull/79252/files#r897301980 I'm not understanding the name here, feels like it should be something like TestBaseDataSparsifier(Generator/Helper/Runner) or something like that and it should have a comment similar to the comment for the base class describing this.

macandro96 replied:
Renamed it to TestDataNormSparsifiersRunner and added comments

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

[ghstack-poisoned]

macandro96 commented on torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:
```diff
@@ -0,0 +1,139 @@
+        padded_data = torch.ones(height + dh, width + dw, dtype=data.dtype, device=data.device)
+        padded_data = padded_data * float('inf')  # can also be replaced with 0 to stop the removal of edge data
+        padded_data[0:height, 0:width] = data
+        unfolded_data = F.unfold(padded_data.unsqueeze(0).unsqueeze(0), kernel_size=sparse_block_shape,
```
done

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164115).

jerryzh168 commented on test/ao/sparsity/test_data_sparsifier.py:
```diff
@@ -235,3 +237,216 @@ def test_nn_embeddings(self):
+                true_sl = true_sl * zpb / sbs[0] / sbs[1]
+                assert sparse_mask.mean() == true_sl
+
+    def run_tests(self):
```
I feel this is structured differently from what we have in other code, for example in https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L702 we just define a TestCase and then some methods in the TestCase class which implements a single test. can we align with that structure?

macandro96 replied:
Do you mean that we convert each function inside the runner class into a separate class and invoke each one of them in the main test case class?

macandro96 replied:
added a TODO comment, will change structure in a later PR

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

Differential Revision: [D37164115](https://our.internmc.facebook.com/intern/diff/D37164115)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164115).

Commit:
Update on "[ao][sparsity] L1 norm based block data sparsifier"


L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Differential Revision: [D37164115](https://our.internmc.facebook.com/intern/diff/D37164115)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164115).
